src/libraries/Native/Unix/System.Native/pal_io.c

```diff
@@ -1405,11 +1409,31 @@ int64_t SystemNative_GetFileSystemType(intptr_t fd)
     // which got deprecated in macOS 10.6, in favor of statfs
     while ((statfsRes = fstatfs(ToFileDescriptor(fd), &statfsArgs)) == -1 && errno == EINTR) ;
     return statfsRes == -1 ? (int64_t)-1 : (int64_t)statfsArgs.f_type;
+#elif defined(TARGET_SUNOS) // we use statvfs (see SystemNative_GetFileSystemTypeAsString) on SunOS
+    (void)fd; // unused
+    return -1;
 #else
     #error "Platform doesn't support fstatfs"
 #endif
 }
 
+char* SystemNative_GetFileSystemTypeAsString(intptr_t fd)
+{
+#if !HAVE_NON_LEGACY_STATFS
+    int statfsRes;
+    struct statvfs statfsArgs;
+    while ((statfsRes = fstatvfs(ToFileDescriptor(fd), &statfsArgs)) == -1 && errno == EINTR) ;
+    return statfsRes == -1 ? NULL : strdup(statfsArgs.f_basetype);
```

Comment:
`strdup` allocates memory that needs to be freed. How are we going to free this memory?

Reply:
This memory doesn't need to be freed explicitly in our code, it will be freed by the GC when managed method goes out of scope. Also, have a look at other `strdup` usages https://github.com/dotnet/runtime/search?q=strdup+path:src/libraries/native.

Reply:
> it will be freed by the GC when managed method goes out of scope

Not quite.  It'll be freed by the interop layer as part of creating a string from the returned pointer.